recipes/full_finetune_single_device.py

```diff
@@ -433,73 +438,44 @@ def _setup_optimizer(
                 model=self._model, optim_dict=optim_dict
             )
             # Create a wrapper for checkpoint save/load of optimizer states when running in backward.
-            self._optim_ckpt_wrapper = training.create_optim_in_bwd_wrapper(
+            optimizer = training.create_optim_in_bwd_wrapper(
                 model=self._model, optim_dict=optim_dict
             )
-            # Load optimizer states. If optimizer states are being restored in an optimizer in backward
-            # run, these need to have been saved with the same setting. Cannot restore from runs that did not
-            # use optimizer in backward.
-            if opt_state_dict is not None:
-                try:
-                    self._optim_ckpt_wrapper.load_state_dict(opt_state_dict)
-                except BaseException as e:
-                    raise RuntimeError(
-                        "Failed loading in-backward optimizer checkpoints."
-                        "Please make sure run being restored from was using in-backward optimizer."
-                    ) from e
-            self._logger.info("In-backward optimizers are set up.")
-            return None
         else:
             optimizer = config.instantiate(cfg_optimizer, self._model.parameters())
 
-            if opt_state_dict:
-                optimizer.load_state_dict(opt_state_dict)
-            self._logger.info("Optimizer is initialized.")
-            return optimizer
+        if opt_state_dict:
+            optimizer.load_state_dict(opt_state_dict)
+
+        self._logger.info("Optimizer is initialized.")
+        return optimizer
 
     def _setup_lr_scheduler(
         self,
-        cfg_lr_scheduler: Optional[DictConfig],
+        cfg_lr_scheduler: DictConfig,
         num_training_steps: int,
         last_epoch: int,
-    ) -> Optional[Optimizer]:
+    ) -> LambdaLR:
```

Comment:
This was embarassing.

Reply:
We should really generalize to generic `LRScheduler`..

Reply:
Talk to me after we add any other LR scheduler besides cosine